### ISSUE TITLE
Support fusing and declarative tiling/distribution of microkernel ops.

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#12300]
(tracking issue for centralizing various discussions - may be a few work items that come out of this as we make progress)

The ultimate goal here is to have `iree_codegen.ukernel.generic` (or some wrapper) at input to iree-compile be formed/fused into dispatch regions as a root op and then tiled/distributed into loops. This would allow for custom user code to be efficiently scheduled in workgroups with generated code, custom target-specific microkernel library matchers (ala AITemplate), and easy experimentation of new codegen strategies.

It's ok if some of the power of something like the TilingInterface is not available in such a form as the goal is to allow dynamic injection of microkernels into an existing iree-compile binary instead of full codegen deeply integrated into the compiler. An attribute that declaratively specified the tiling information in order to service TilingInterface would allow users to express how to perform the basic tiling of the microkernel and the interface implementation could sample that when available and otherwise return defaults (one tile for the entire problem as a single call, no workgroup distribution).

Expecting something like this passed to iree-compile:
```mlir
%in0 = mhlo.add ... : tensor<4096xf32>
%out0 = tensor.empty : tensor<4096xf32>
%0 = iree_codegen.ukernel.generic "foo" ins(%in0 : tensor<4096xf32>) outs(%out0 : tensor<4096xf32>) {
  objects = #hal.executable.objects<...> defining where to source the ukernel code,
  some attributes to control tiling,
}
```
A `custom_dispatch` sample should be able to have the above code and run end-to-end. Today the user needs to form the flow.executable, tile and distribute manually, and then add the call to the ukernel op.

I believe the major places that we need to pass through are dispatch region formation and tile-and-distribute. FormDispatchRegions has isRootOp return true if the op implements TilingInterface. TileAndDistributeToWorkgroups also uses the TilingInterface. I'm not well-versed in the pipeline though and there may be some other things we need to do but at least implementing the TilingInterface seems required. I'm not sure where workgroup count is calculated but I'm guessing it routes through the same places.

My naïve survey shows these TilingInterface methods being used in TileDispatchUsingInterface:
- getLoopIteratorTypes: could be IteratorTypeArrayAttr?
- getIterationDomain: could be from indexing maps?
- getResultTilePosition: could be from indexing maps?
- getTiledImplementation: could be from indexing maps?

I'm not familiar with linalg.generic's library call feature but with LinalgStructuredInterface it requires a lot of stuff that would be difficult to declaratively specify and isn't really needed for this purpose (we aren't trying to transform microkernels, just tile them). I also like the idea of having a stable IREE op we can present that is not attached to linalg and ukernel.generic seems ideal (even if we create an iree_input dialect variant we use for insulation). Effectively we'd be getting the ability to do things like LinalgNamedStructuredOps.yaml but in attribute form and instead of defining the body we'd define the external call and where to source the ukernel objects from. Unlike the named ops since all information is carried hermetically on the op itself we can go from python all the way through to runtime without changing the compiler or runtime.

Since ukernels have a much more rigid format we can simplify things as much as we need to make things tractable. To start we could have it be some basic "tile on operand 1 dimensions 1,2" or whatever and infer the rest and we may always want something simple like that. I don't know enough to do that but maybe @bjacob could help given that he's got the best idea of how to tile microkernels?

Relates: #12292 (could build cutlass templating at the frontend), #12222 (enables experimentation without compiler changes), somewhat #12283 (way to define custom kernels out of tree for experimentation).

**Comment by benvanik** [comment]
Had some good brainstorming with @bjacob and @MaheshRavishankar, both can probably articulate the next steps better than me but I think I understand more now than I did this morning and can at least refine my above request to be less hand-wavy and provide some motivation :)

**TLDR** is we need a new (probably HAL) device call op that carries some declarative scheduling information or ways to get it (other device calls for tile size, etc) and plumbing added to get it through dispatch region formation and tile and distribute. Next steps are sketching out what scheduling information is needed to satisfy fusion/tiling/etc and how we might represent it.

Please call out anything that sounds intangible/intractable and I'd love to hop on a gvc and brainstorm more!

---

Mahesh's #12157 is shining a light on a new path by demonstrating how microkernels are just device functions that are called within a workgroup (and possibly nested in a loop within that workgroup) to act on a chunk of data. This holds on all backends: you can call a `__device__` function from another `__device__` function in CUDA, or make SPIR-V calls, or have language-level calls even if only able to bake them out as inlined code. Where the external function comes from can vary - it could be in the same compilation unit, an external module, or linked in any number of ways - but it's just a function. IREE is pretty much just a device code linker like any other with some fancy multi-targeting capabilities and we're in the same realm of capability as clang/ld/nvcc/etc (https://developer.nvidia.com/blog/separate-compilation-linking-cuda-device-code/, https://github.com/KhronosGroup/SPIRV-Tools/blob/main/tools/link/linker.cpp, etc) from a mechanism perspective.

What really makes a microkernel a "microkernel" instead of just a "device function call" is defining the API signature of the function, how to call that function in loops within a workgroup, and how to distribute those loops to multiple workgroups. That is to say the microkernel has a _schedule_ defining how it is to be called but the _mechanism_ in which it's called is always going to be a function call.

If we split the concept then of microkernels into the _schedule_ and the _mechanism_ the mechanism is the easy part. We've now got several ways to handle functions (and can add many more) just as any compiler/linker does: a target function can be defined in MLIR right beside a call to it, in an external MLIR file we load and merge into the module (effectively the same), in an LLVM IR bitcode file or SPIR-V module produced from any language and tool that can produce them (where we can still inline/do fun cross-call stuff), in a machine code object file (.o, .a, .ptx), or in an external dynamically linked library (.so, imports injected at runtime, VM modules). The IR modulo the _schedule_ is going to look the same on any platform just as a function call looks the same in C, CUDA C++, GLSL/HLSL, etc even if the syntax differs a bit. All IREE device targets are expected to support function calls at some level (even if only in MLIR) and our higher layers of the stack can assume a `func.call` is always going to be lowerable even if only when in the same module and inlined.

---

To make this more concrete lets work an example from the bottom up: here's the mmt4d microkernel as just an extern function declaration and a call to it within a loop inside of a workgroup (**note this doesn't include dynamic tile sizes yet, but that's definitely part of this**):
```mlir
func.func private @vmvx.matmul.i8i8i32(memref<i8>, index, index, memref<i8>, index, index, memref<i32>, index, index, index, index, index, i32) attributes {vm.import.module = "vmvx"}
func.func @matmul_i8_dispatch_0_matmul_DxDxD() {
  // ...
  %11 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_y]
  %12 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_y]
  scf.for %arg0 = %11 to %4 step %12 {
    %13 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg0)[%4]
    %14 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
    %15 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
    scf.for %arg1 = %14 to %6 step %15 {
      %16 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg1)[%6]
      %17 = affine.apply affine_map<()[s0, s1] -> (s0 * s1)>()[%arg0, %5]
      %18 = affine.apply affine_map<()[s0, s1, s2] -> (s0 * s1 + s2)>()[%arg0, %7, %arg1]
      func.call @vmvx.matmul.i8i8i32(%8, %17, %5, %9, %arg1, %6, %10, %18, %7, %13, %16, %16, %c0_i32) : (memref<i8>, index, index, memref<i8>, index, index, memref<i32>, index, index, index, index, index, i32) -> ()
    }
  }
  return
}
```

This function call was mechanically converted from a bufferized `iree_codegen.ukernel.generic`:
```mlir
  %15 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_y]
  %16 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_y]
  scf.for %arg0 = %15 to %6 step %16 {
    %17 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg0)[%6]
    %18 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
    %19 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
    scf.for %arg1 = %18 to %9 step %19 {
      %20 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg1)[%9]
      %subview = memref.subview %14[%arg0, %arg1] [%17, %20] [1, 1] : memref<?x?xi32> to memref<?x?xi32, strided<[?, 1], offset: ?>>
      %subview_0 = memref.subview %12[%arg0, 0] [%17, %7] [1, 1] : memref<?x?xi8> to memref<?x?xi8, strided<[?, 1], offset: ?>>
      %subview_1 = memref.subview %13[0, %arg1] [%7, %20] [1, 1] : memref<?x?xi8> to memref<?x?xi8, strided<[?, 1], offset: ?>>
      iree_codegen.ukernel.generic "vmvx.matmul.i8i8i32" ins(%subview_0, %subview_1 : memref<?x?xi8, strided<[?, 1], offset: ?>>, memref<?x?xi8, strided<[?, 1], offset: ?>>) outs(%subview : memref<?x?xi32, strided<[?, 1], offset: ?>>) (%17, %20, %20, %c0_i32 : index, index, index, i32)
    }
  }
```

Which came from a tensor form of the ukernel op using the destination passing style interface:
```mlir
    %15 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_y]
    %16 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_y]
    scf.for %arg0 = %15 to %6 step %16 {
      %17 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg0)[%6]
      %18 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
      %19 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
      scf.for %arg1 = %18 to %9 step %19 {
        %20 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg1)[%9]
        %21 = flow.dispatch.tensor.load %14, offsets = [%arg0, %arg1], sizes = [%17, %20], strides = [1, 1] : !flow.dispatch.tensor<writeonly:tensor<?x?xi32>>{%10, %11} -> tensor<?x?xi32>
        %22 = flow.dispatch.tensor.load %12, offsets = [%arg0, 0], sizes = [%17, %7], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<?x?xi8>>{%6, %7} -> tensor<?x?xi8>
        %23 = flow.dispatch.tensor.load %13, offsets = [0, %arg1], sizes = [%7, %20], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<?x?xi8>>{%8, %9} -> tensor<?x?xi8>
        %dim = tensor.dim %22, %c0 : tensor<?x?xi8>
        %dim_0 = tensor.dim %23, %c1 : tensor<?x?xi8>
        %dim_1 = tensor.dim %21, %c1 : tensor<?x?xi32>
        %24 = iree_codegen.ukernel.generic "vmvx.matmul.i8i8i32" ins(%22, %23 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%21 : tensor<?x?xi32>) (%dim, %dim_0, %dim_1, %c0_i32 : index, index, index, i32) -> tensor<?x?xi32>
        flow.dispatch.tensor.store %24, %14, offsets = [%arg0, %arg1], sizes = [%17, %20], strides = [1, 1] : tensor<?x?xi32> -> !flow.dispatch.tensor<writeonly:tensor<?x?xi32>>{%10, %11}
      }
    }
```

Which came from a pattern match of a single tile of a linalg.matmul:
```mlir
      %15 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_y]
      %16 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_y]
      scf.for %arg0 = %15 to %6 step %16 {
        %17 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg0)[%6]
        %18 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_id_x]
        %19 = affine.apply affine_map<()[s0] -> (s0 * 64)>()[%workgroup_count_x]
        scf.for %arg1 = %18 to %9 step %19 {
          %20 = affine.min affine_map<(d0)[s0] -> (-d0 + s0, 64)>(%arg1)[%9]
          %21 = flow.dispatch.tensor.load %12, offsets = [%arg0, 0], sizes = [%17, %7], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<?x?xi8>>{%6, %7} -> tensor<?x?xi8>
          %22 = flow.dispatch.tensor.load %13, offsets = [0, %arg1], sizes = [%7, %20], strides = [1, 1] : !flow.dispatch.tensor<readonly:tensor<?x?xi8>>{%8, %9} -> tensor<?x?xi8>
          %23 = tensor.empty(%17, %20) : tensor<?x?xi32>
          %24 = linalg.fill ins(%c0_i32 : i32) outs(%23 : tensor<?x?xi32>) -> tensor<?x?xi32>
          %25 = linalg.matmul {lowering_config = #iree_codegen.lowering_config<tile_sizes = [[64, 64, 0]]>} ins(%21, %22 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%24 : tensor<?x?xi32>) -> tensor<?x?xi32>
          %26 = arith.index_castui %4 : i32 to index
          %27 = arith.index_castui %5 : i32 to index
          flow.dispatch.tensor.store %25, %14, offsets = [%arg0, %arg1], sizes = [%17, %20], strides = [1, 1] : tensor<?x?xi32> -> !flow.dispatch.tensor<writeonly:tensor<?x?xi32>>{%26, %27}
        }
      }
```

Which came from an untiled linalg.matmul within a dispatch:
```mlir
  %9 = tensor.empty(%6, %7) : tensor<?x?xi32>
  %10 = linalg.fill ins(%c0_i32 : i32) outs(%9 : tensor<?x?xi32>) -> tensor<?x?xi32>
  %11 = affine.apply affine_map<()[s0, s1, s2] -> ((s1 - s0) ceildiv s2)>()[%c0, %0, %c1]
  %12 = affine.apply affine_map<()[s0, s1, s2] -> ((s1 - s0) ceildiv s2)>()[%c0, %4, %c1]
  %13 = flow.dispatch.region[%11, %12, %c1] -> (tensor<?x?xi32>{%6, %7}) {
    %23 = linalg.matmul ins(%2, %5 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%10 : tensor<?x?xi32>) -> tensor<?x?xi32>
    flow.return %23 : tensor<?x?xi32>
  } count(%arg3: index, %arg4: index, %arg5: index) -> (index, index, index) {
    %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg3, %arg4, %arg5
    flow.return %x, %y, %z : index, index, index
  }
```

Which came from a full-tensor linalg.matmul at the input layer:
```mlir
%0 = linalg.matmul ins(%arg0, %arg1 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%arg2 : tensor<?x?xi32>) -> tensor<?x?xi32>
```

---

So then the question becomes: if the schedule is the unique thing about a microkernel and linalg.matmul can make it down to a device function call by using the linalg.matmul op to derive the schedule _can we express that schedule in such a way that we can have that microkernel device function call earlier in the pipeline without needing the linalg.matmul_? e.g. starting with something like this (**but not this**) and dropping the marble down to the first IR snippet that has just the device function call in a loop nest within a workgroup:
```mlir
%0 = iree_codegen.ukernel.generic "vmvx.matmul.i8i8i32" ins(%arg0, %arg1: tensor<?x?xi8>, tensor<?x?xi8>) outs(%arg2 : tensor<?x?xi32>) (%dim, %dim_0, %dim_1, %c0_i32 : index, index, index, i32) -> tensor<?x?xi32> {
  // attributes controlling tiling ala #iree_codegen.lowering_config<tile_sizes = [[64, 64, 0]]>
}
```

The interesting leap here and what this issue is reaching for is if we can express "`func.call @some_device_function` with this distribution and this tiling schedule" such that the iree-compile binary has no idea what `@some_device_function` is but is still able to get to the final form we'd have allowed the ecosystem to define arbitrary microkernels at any layer of the stack! _And_ while still able to get all the goodness we can deliver with multi-targeting, fusion, and concurrency/parallelization without needing to solve the whole program loop expansion and tiling problem that is at odds with our core requirement of retargetability. An end user with some custom algorithms they need to run on device (image processing, data structure lookup, etc) can now seamlessly mix that with the rest of their model that we process - tosa/mhlo ops side-by-side with their custom ops without giving up fusion/tiling/etc and getting a consistent mechanism across all their targets. A hardware vendor can provide a library of microkernels that is processed after we form dispatch regions as part of executable preprocessing in #12313 and an end-user (or bundler) could package up several such libraries as part of a solution. An ML deployment engineer needing to meet a deadline could analyze their bottleneck and quickly hack together a specific solution for their model that they can inject right from the python layer ala triton but while retaining IREE's AOT and multi-targeting deployment capabilities. etc. etc. How to pattern match and substitute ops, or how to represent such ops in jax, etc is out of scope but an interesting area to explore that builds on top of this - transform dialect, python scripts using the MLIR API to do whatever the user wants, etc. Today people basically have to become an MLIR expert and land code in the main IREE repo to do this kind of stuff and this would be the first big step (and arguably the most critical) towards fixing that.

Note that this doesn't mean we want all end users to be doing this ("I just want to run my tflite model, wtf is all this?") but that we want to enable an ecosystem and extensible compiler stack that allows for the same flexibility as in normal compilers ("I'll use the STL and libjpeg and write half the program in rust to build my executable") but with the added stuff unique to what we can provide. It also doesn't mean that we don't want microkernels in the core compiler and instead this just allows us to be very selective about which we allow in: llvm recognizes memset and memcpy as intrinsics but not atoi or DecodeMyJPEGFile and that's what we're going for. iree-compile shifts to support a common subset of generally applicable and sustainably maintainable microkernels for the 90% of the perf that matters (conv/gemm/etc), openxla/shark/etc frameworks integrating/whitelabeling IREE can mix in some of their own including vendor-specific libraries outside the scope of what we provide in the core, and users can decide what mix of things to use themselves as needed including their own custom stuff without needing to become a compiler engineer.

Credit to linalg for having had library calls since almost the beginning and this is somewhat related but the linalg approach has required structured ops in the compiler (even if generated from yaml and such) putting it into compiler engineering territory. Until the TilingInterface and destination passing style reworking it also wasn't clear how to make them work with our requirements around multi-targeting and host/device partitioning. Maybe when done there's a trivial lowering from linalg ops with library calls specified to this new stuff that lets them be passed through IREE when doing target-specific stuff but it's not a mechanism intended for the scope of what we're doing and I don't think there's much overlap.

---

All that said, I believe we're really needing some op that satisfies "`func.call @some_device_function` with this distribution and this tiling schedule" - `iree_codegen.ukernel.generic` isn't it but it's closer than we've ever been, and a rough sketch would be something like this implementing TilingInterface and a few other things:
```mlir
%... = hal.device.call "some_name" ins(...) outs(...) args(...) -> ... {
   // attributes or regions controlling fusion/tiling/etc using affine maps and stuff
   objects = #hal.executable.objects<{
     #hal.executable.target<"llvm-cpu", "embedded-elf-arm_64"> = [
       #hal.executable.object<{path = "some/file_arm_64.o"}>
     ],
     #hal.executable.target<"cuda", "cuda-nvptx-fb", {target_arch = "sm_35"}> = [
       #hal.executable.object<{path = "some/file_cuda_sm_35.bc"}>
     ],
     #hal.executable.target<"vmvx", "..."> = [
       #hal.executable.object<{path = "builtin:vmvx"}>
     ]
   }>
}
```

If we could have that at the top level in tensor form we could lower it into `func.func` + `func.call` as Mahesh has `iree_codegen.ukernel.generic` wired up in #12157. The critical simplifying requirement is that the op is hermetic and any regions are isolated from above - that allows us to trivially move them around the program, move them into dispatch regions, outline them into executables, and do whatever mangling we want during executable translation. We'd peel off the linkage information automatically and could even use the same style of target -> value mappings for the other attributes if needed.

For many ops attributes are likely enough to express what we need - "use these dimensions of these operands/results for the iteration space" and "these are the loop iterator types" and "here's a set of affine maps to do ... affine map stuff" - and we can always make it more complex as needed. One example would be dynamic tile size calculation: instead of hardcoding the tile size we could either have a region that computes it based on a `!hal.device` just like our workgroup count region does and/or reference _another device function_. For example this which would use a hypothetical `tile_size_call` attribute to emit a device-side `func.call @mmt4d_query_tile_size` that whatever is providing the definition of `mmt4d` _also_ provides, preserving multi-targeting and configuration-specific specialization:
```mlir
%... = hal.device.call "mmt4d" ins(...) outs(...) args(...) -> ... {
   tile_size_call = "mmt4d_query_tile_size"
}
```

I think the biggest questions are around the ergonomics and initial way of satisfying the fusion/tiling interface requirements of this new op. I have a handle on defining it and plumbing it through flow/stream/hal layers and the linkers but am not familiar enough to know how to define the attributes we need for enabling dispatch region formation or the tiling part during TileAndDistributeToWorkgroups and `workgroup_count_from_dag_root`. It'd be useful to sketch out here on the issue what people would want to express using mmt4d and flash attention/other linalg_ext ops as examples.

There's follow-on work to have a versioned op in the `iree_input` dialect for easier frontend use, but we can get a tremendous amount of mileage out something at the `hal` dialect level today (as just as we have `hal.tensor.import` and such we can have certain HAL ops all the way at the frontend).

Wow this got long.
![image](https://user-images.githubusercontent.com/75337/220846975-bb9e1cb5-af45-4c60-9d93-3012c981a47a.png)

**Comment by stellaraccident** [comment]
User level manifestations of this could be very interesting. If we ended up with some kind of `tiled_device_func` and `tiled_device_call` at the top level, and if that were general -- that would be a pretty accessible way for the frameworks to interop.

**Comment by stellaraccident** [comment]
Or multiple funcs with a single call. The one mentioned would be inline with the module and defined with high level ops just like any other function. But there could also be a different func for providing device blobs.

**Comment by stellaraccident** [comment]
It'd be great if the user manifestation was more "functions" and less "deep region hierarchies" in my personal opinion.

**Comment by stellaraccident** [comment]
To level set, I've seen this observation multiple times but I've never seen one hit escape velocity and not just be a niche thing of some backend. I judge this to typically be because it was never connected all the way up in a user serviceable way that composed well (or at all). We inherit some advantages there from MLIR existing from top to bottom and if it can be formulated right, this could make it where previous iterations didn't.

**Comment by benvanik** [comment]
Yeah! I've been toying with how this connects upwards for larger scopes and it's likely going to look a lot like CUDA's host/device/global annotations (but something we could in many cases infer ourselves). The above is trying to solve for the intrinsic-style stuff (functions not in the source program/visible compilation units) and the next step upwards in the layering would be to allow just normal functional control flow. We've already got that fidelity one step after dispatch region formation in `flow.executable`: it's an entire module and can have as many functions, calls, etc it wants and this is how our custom dispatch sample works as if passing in an executable like input from the frontend: https://github.com/openxla/iree/blob/309dc5b1f7ca6093fcf1b808cfecc82c5786d632/samples/custom_dispatch/cpu/embedded/example_stream.mlir#L110-L178

If you squint our compiler looks effectively just like CUDA's stack (with obvious differences, squint hard):
![image](https://user-images.githubusercontent.com/75337/220966149-facda6e8-9a70-4dfb-a63b-fa4c8e4cfda9.png)
(it's not by chance - how CUDA's stack is put together is pretty much the best way to do it for something like C++ - we can do better though without having those language-defined programming model limitations)

Dispatch region formation is effectively our way of partitioning host/device from the .cu automatically in structured forms, but it's also possible to partition much more directly and also do it explicitly ala `__device__` / `__host__` in CUDA. Once we have the above stuff solved then really we are talking about how to do that original .cu processing to slice out which side of the stack wants what.

As an example of an incremental change that moves upward say we allowed calls inside of dispatch regions and then enhanced the OutlineDispatchRegionsPass so we could take something like this and turn it into the executable (always the intent, never done):
```mlir
func.func private @some_utility_function(...) {
  whatever
}
...
  %13 = flow.dispatch.region[%11, %12, %c1] -> (tensor<?x?xi32>{%6, %7}) {
    %23 = linalg.matmul ins(%2, %5 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%10 : tensor<?x?xi32>) -> tensor<?x?xi32>
    %24 = func.call @some_utility_function(%23)
    flow.return %24 : tensor<?x?xi32>
  } count(%arg3: index, %arg4: index, %arg5: index) -> (index, index, index) {
    %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg3, %arg4, %arg5
    flow.return %x, %y, %z : index, index, index
  }
```
->
```mlir
flow.executable @exe {
  flow.executable.export @dispatch
  builtin.module {
    func.func private @some_utility_function(...) {
      whatever
    }
    func.func @dispatch(...) {
      linalg.matmul
      func.call @some_utility_function
    }
  }
}
+
flow.dispatch @exe::@dispatch
```

For simple functions not taking tensors/etc that'd be pretty easy to do today - we'd scan the call tree and ensure its pure and just clone out the slice. Of course we could get fancier here and do things like produce a `flow.executable` of just the library functions and then link that to all executables, or link earlier in the pipeline such that all the `flow.executables` get merged (today we don't because we parallelize translation on the executables, but there's potential fixes for that). But mechanically it'd work.

Moving one step above that would be to form the dispatch regions with the calls in them:
```mlir
%23 = linalg.matmul ins(%2, %5 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%10 : tensor<?x?xi32>) -> tensor<?x?xi32>
%24 = func.call @some_utility_function(%23)
```
->
```mlir
  %13 = flow.dispatch.region[%11, %12, %c1] -> (tensor<?x?xi32>{%6, %7}) {
    %23 = linalg.matmul ins(%2, %5 : tensor<?x?xi8>, tensor<?x?xi8>) outs(%10 : tensor<?x?xi32>) -> tensor<?x?xi32>
    %24 = func.call @some_utility_function(%23)
    flow.return %24 : tensor<?x?xi32>
  } count(%arg3: index, %arg4: index, %arg5: index) -> (index, index, index) {
    %x, %y, %z = flow.dispatch.workgroup_count_from_dag_root %arg3, %arg4, %arg5
    flow.return %x, %y, %z : index, index, index
  }
```

This is where having the declarative ways of defining fusing/tiling would be useful. It then comes down to `func.func` being insufficient for expressing things we need and that calls currently break tiling and bufferization once inside. Inlining within each executable would help the latter but the former is likely calling for a new func op. I've long wanted a `util.func`/`util.call` that supported the TiedOpInterface and once we have that we could implement whatever interfaces and declarative tiling form we end up with for the intrinsic style `hal.device.call` and then we wouldn't have to force inlining, could still tile/distribute within the executable, and still bufferize (with work on bufferization).

With the right hacks one could actually solve it with `func.func` - register the external interfaces we need (TiedOpInterface, TilingInterface, etc) on `func.call` and have the external models query attributes on the function. A step that ran through and propagated from `func.func` -> `func.call` so that's efficient is probably called for, at which point just having our own ops is worth it to avoid the duck typing.

So yeah, strong agreement on the direction - I think as we continue bumping up the layering here so that executable translation accepts more, tiling accepts more, and dispatch region formation accepts more we end up at a place where we have both the user-apparent ability to mix stuff freely while also retaining all of the cool stuff we can do. That's exciting :)

**Comment by bjacob** [comment]
Discussed here: https://discord.com/channels/689900678990135345/689932666505920596/1078404559757647993

(summary: start incremental, don't try to solve dynamic data tiling yet, maybe even reuse lowering_config if expedient?)

**Comment by MaheshRavishankar** [comment]
I am late to the party, cause I wanted to digest some of this. I might not have digested all of this, but its a really good summary from Ben above. I think I will just restate most of it in my words, and then compare notes with what Ben has above.

Lets start with a very simple operation definition

```
%0:2 = iree_input.tilable.op tiled_implementation_fn "tile_implementation_foo"
    ins(%in1, %in2, ... tensor<..> ,tensor<..>, ...) outs(%out1, %out2, ... : tensor<..>, tensor<..>, ...)
    (%other_operands1, %other_operands2, ....) -> tensor<..>, tensor<..>, ...
```

This operation is very similar to a `linalg` op. It is not by accident. I think Linalg ops were meant to have the functionality being described here for a single operation. It seems to be another case of a Linalg concept being decoupled from Linalg op/dialect to expose just that feature and not inherit all the other constraints of a Linalg op. `ins` represents the inputs of the operation, and `outs` the outputs. The `tiled_implementation_fn` provides the name of the micro kernel. 

As it stands today there are two interfaces that are currently needed that this operation needs to implement to enable it to be compiled down to actually invoking the microkernel

1) `TilingInterface` : This allows this operation to be moved into its own dispatch, and do one level of tile and distribute. As written today, if all methods of the interface are implemented the operation can be tiled and distributed, and can also fuse with its consumers. To allow fusing with producers, some enhancement to the TilingInterface are needed (have been bouncing around some ideas, but not fully flushed out to talk about yet).
2) `DestinationPassingStyleOpInterface` : This based on the operation definition is exactly how Linalg defines this interface.

The problem now is how to use the operation definition to automatically implement the `TilingInterface`. One way is to add couple of extra-attributes that could help
```
- iterator_types : StrArrayAttr
- indexing_maps :  AffineMapArrayAttr
```
We know how to use these to implement tiling cause we do those for Linalg operations. With these two attributes you can tile and distribute (and actually with these attributes even fusion with producers/consumers get enabled :P ). Side Note: This operation is looking very similar to a Linalg op but with few constraints dropped.
```
%0 = scf.for ...
  %1 = scf.for ...
    ....
    %2 = iree_codegen.ukernel.generic "tile_implementation_foo" ins(%tiled_input1, %tiled_input2, ... : tensor<..>, tensor<..>)
        outs(%tiled_out1, %tiled_out2, ... : tensor<...>, tensor<...>) (%other_operands1, %other_operands2, ....) 
        -> tensor<...>, tensor<...>, ....
```
which can then bufferize in place to

```
scf.for ...
  scf.for ...
    ....
    iree_codegen.ukernel.generic "tile_implementation_foo" ins(%tiled_input1, %tiled_input2, ... : memref<..>, memref<..>)
        outs(%tiled_out1, %tiled_out2, ... : memref<...>, memref<...>) (%other_operands1, %other_operands2, ....) 
```

which then connects to the existing micro-kernel work.

The remaining question is how do you get the tile sizes for the distribution, I think we could add another field to the op definition
```
iree_input.tilableop .... tile_size_fn "tile_size_fn_name" ....
```
which could just be the name of the function that can be called to get the tile sizes. If we want to just start with static tile sizes this could be just a list of index values.
```
iree_input.tilableop .... tile_sizes [10, 20, ...]
```
This seems like a viable path to me to start prototyping something. Would be really good to have a concrete use case though. I am not sure we should design this in a vaccum .

**Comment by allieculp** [comment]
@MaheshRavishankar @benvanik @bjacob @stellaraccident This one got a bit stale, is it a priority at this time or P2/3?

**Comment by MaheshRavishankar** [comment]
Parts of this is landing. This is more a "we would like to get here" kind of thing. Maybe P2?
